# Transfer learning 

**Transfer learning** is a research problem in machine learning that focuses on storing knowledge gained while solving one problem and applying it to a different but related problem.

In computer vision this means that we can use many open sourced deep learning models which were trained on a huge amounts of images and took a lot of time to train. 

Tensorflow makes it very easy to use powerful CV models and tailor them to your specific problem. 

# Sign language data 

![](sign_language_project/pictures/amer_sign2.png)

The American Sign Language letter database of hand gestures represent a multi-class problem with 24 classes of letters (excluding J and Z which require motion). 

The original dataset is obtain from here: 

https://www.kaggle.com/datamunge/sign-language-mnist 

The images are in a .csv file where the first element of a row is the label and the rest are the pixel values. The original image dimensions are 28x28x1. 

# Reading the data

In [19]:
import os 
import numpy 
import csv
import numpy as np

def get_data(filename: str):
    """
    A function to read the data given a filename
    """
    images = []
    labels = []
    with open(filename) as csvfile:
        for i, row in enumerate(csvfile):
            # Skipping the header
            if i > 0:
                all_val = row.split(',')

                label = int(all_val[0])
                pixels = np.asarray(all_val[1:]).astype(float).reshape(28, 28)

                labels.append(label)
                images.append(pixels)
    
    # Converting to arrays
    images = np.asarray(images)
    labels = np.asarray(labels)
    
    return images, labels

path_sign_mnist_train = f"{os.getcwd()}/sign_language_project/data/sign_mnist_train.csv"
path_sign_mnist_test = f"{os.getcwd()}/sign_language_project/data/sign_mnist_test.csv"
training_images, training_labels = get_data(path_sign_mnist_train)
testing_images, testing_labels = get_data(path_sign_mnist_test)

# Dimensions of data
print(training_images.shape)
print(training_labels.shape)
print(testing_images.shape)
print(testing_labels.shape)

(27455, 28, 28)
(27455,)
(7172, 28, 28)
(7172,)


There are 27455 training images and 7172 test images.

## Loading the model

In [14]:
# The bellow code loads the model architecture
from tensorflow.keras.applications.resnet50 import ResNet50

# Downloading the weights
# The include_top=False indicates that we will using our custom input layer
model = ResNet50(include_top=False, weights='imagenet')

94773248/94765736 [==============================] - 20s 0us/step


In [12]:
# Number of layers
print(len(model.layers))

177


The model is huge, containing 177 layers and millions of unique weights. One thing to note is that the input dimensions for the model are 224x224x3.

The model outputs probabilities regarding 1000 classes. This will not be the case for our data thus we will freeze the top n layers of the model, add some custom layers at the bottom of the model and train it with our data.

# Transfering learning

In [ ]:
We first need to decide which layer to 